src/platform/unix/mod.rs

```diff
@@ -415,18 +432,20 @@ impl OsIpcChannel {
 
 pub struct OsIpcReceiverSet {
     incrementor: Incrementor,
-    pollfds: Vec<pollfd>,
-    fdids: HashMap<c_int, u64, BuildHasherDefault<FnvHasher>>
+    poll: Poll,
+    pollfds: HashMap<Token, FDHandle, BuildHasherDefault<FnvHasher>>,
+    events: Events
 }
 
 impl Drop for OsIpcReceiverSet {
     fn drop(&mut self) {
-        unsafe {
-            for pollfd in self.pollfds.iter() {
-                let result = libc::close(pollfd.fd);
-                assert!(thread::panicking() || result == 0);
+        let mut result = 0;
+        for (_, &FDHandle{ id: _, descriptor: fd }) in &self.pollfds {
+            unsafe {
+                result |= libc::close(fd);
```

Comment:
I'm not sure how moving the assert out of the loop corresponds to the quoted passage?...

I realised though that going through the entire loop first might indeed be beneficial: this way if closing one descriptor fails, we don't leave the others leaking before triggering the thread panic... However, since this is entirely unrelated to the `mio` change as far as I can tell, I'd say this should better go into a separate PR?

Anyway, looking at the newest variant of this code, I'm beginning to wonder whether we are somehow talking past each other here... You now don't try to close at all during panic? That's definitely not right.

(Also, even though Rust asserts are not optional like in C, I still don't think it's a good idea for them to have side effects -- both for readability of the code, and of the panic messages...)

I really believe it would be best if you just replace the `Vec` with `HashMap`, and otherwise leave the code unchanged here... As a general rule of thumb, I think functional changes should be kept minimal; while cleanups and other unrelated improvements better go into separate commits -- or even separate PRs, if the changes in question have a major effect, or might be controversial in other ways...